libyul/backends/evm/OptimizedEVMCodeTransform.cpp

```diff
@@ -0,0 +1,431 @@
+		}, _slot);
+	};
+
+	yulAssert(m_assembly.stackHeight() == static_cast<int>(m_stack.size()), "");
+	// ::createStack asserts that it has achieved the target layout.
+	::createStackLayout(m_stack, _targetStack | ranges::to<Stack>, [&](unsigned _i) {
+		yulAssert(_i > 0, "");
+		if (_i > 16)
+		{
+			int deficit = static_cast<int>(_i - 16);
+			StackSlot const& deepSlot = m_stack.at(m_stack.size() - _i - 1);
+			YulString varNameDeep = slotVariableName(deepSlot);
+			YulString varNameTop = slotVariableName(m_stack.back());
+			string msg = "Cannot swap " + (varNameDeep.empty() ? "Slot " + stackSlotToString(deepSlot) : "Variable " + varNameDeep.str()) +
+				" with " + (varNameTop.empty() ? "Slot " + stackSlotToString(m_stack.back()) : "Variable " + varNameTop.str()) +
+				": too deep in the stack by " + to_string(deficit) + " slots in " + stackToString(m_stack);
+			m_stackErrors.emplace_back(StackTooDeepError(
+				m_currentFunctionInfo ? m_currentFunctionInfo->function.name : YulString{},
+				varNameDeep.empty() ? varNameTop : varNameDeep,
+				deficit,
+				msg
+			));
+
+			m_assembly.markAsInvalid();
+		}
+		else
+			m_assembly.appendInstruction(evmasm::swapInstruction(_i));
+	}, [&](StackSlot const& _slot) {
+		if (auto depth = util::findOffset(m_stack | ranges::views::reverse, _slot))
+		{
+			if (*depth < 16)
+			{
+				m_assembly.appendInstruction(evmasm::dupInstruction(static_cast<unsigned>(*depth + 1)));
+				return;
+			}
+			else if (!canBeFreelyGenerated(_slot))
+			{
+				int deficit = static_cast<int>(*depth - 15);
+				YulString varName = slotVariableName(_slot);
+				string msg = (varName.empty() ? "Slot " + stackSlotToString(_slot) : "Variable " + varName.str())
+					+ " is " + to_string(*depth - 15) + " too deep in the stack " + stackToString(m_stack);
+				m_stackErrors.emplace_back(StackTooDeepError(
+					m_currentFunctionInfo ? m_currentFunctionInfo->function.name : YulString{},
+					varName,
+					deficit,
+					msg
+				));
+				m_assembly.markAsInvalid();
+				m_assembly.appendConstant(u256(0xDEADBEEF));
+				return;
+			}
+		}
+		yulAssert(static_cast<int>(m_stack.size()) == m_assembly.stackHeight(), "");
+		std::visit(util::GenericVisitor{
+			[&](LiteralSlot const& _literal)
+			{
+				m_assembly.setSourceLocation(locationOf(_literal));
+				m_assembly.appendConstant(_literal.value);
+			},
+			[&](FunctionReturnLabelSlot const&)
+			{
+				yulAssert(false, "Cannot produce function return label.");
+			},
+			[&](FunctionCallReturnLabelSlot const& _returnLabel)
+			{
+				if (!m_returnLabels.count(&_returnLabel.call.get()))
+					m_returnLabels[&_returnLabel.call.get()] = m_assembly.newLabelId();
+				m_assembly.appendLabelReference(m_returnLabels.at(&_returnLabel.call.get()));
+			},
+			[&](VariableSlot const& _variable)
+			{
+				if (m_currentFunctionInfo)
+					if (util::contains(m_currentFunctionInfo->returnVariables, _variable))
+					{
+						m_assembly.appendConstant(0);
+						return;
```

Comment:
This is a blind spot in the verification, I guess. In practice, this will only happen on the first use of the return variable and the stack layout generator will propagate any use e.g. through loops, so it should be fine, but I don't see an easy way to verify that in here.

Reply:
But it might warrant some dedicated test cases at least.